File: Data Structures & Algorithms/count-number-of-islands/submission-2.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]: return 0


        ans = 0
        ROWS, COLS = len(grid), len(grid[0])

        def dfs(r, c):
            dirs = [[1,0],[0,1],[-1,0],[0,-1]]

            if (r < 0 or r >= ROWS or c < 0 or c >= COLS or grid[r][c] == '0'): return

            grid[r][c] = '0'
            for nr, nc in dirs:
                dfs(r + nr, c + nc)

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == '1':
                    ans += 1
                    dfs(r, c)
        return ans

File: Data Structures & Algorithms/islands-and-treasure/submission-2.cpp
class Solution {
public:
    void islandsAndTreasure(vector<vector<int>>& grid) {
        if (grid.empty() || grid[0].empty()) return;
        queue<pair<int,int>> q;
        vector<pair<int,int>> dirs = {{1,0},{0,1},{-1,0},{0,-1}};
        int INF = 2147483647;
        int ROWS = grid.size(), COLS = grid[0].size();

        //seed the q
        for (int r = 0; r < ROWS; r++){
            for (int c = 0; c < COLS; c++){
                if (grid[r][c] == 0) q.push({r,c});
            }
        }

        while (q.size()){
            int row = q.front().first, col = q.front().second;
            q.pop();

            for (auto& [dr, dc] : dirs){
                int newR = dr + row, newC = dc + col;
                if (min(newR, newC) < 0 || newR == ROWS || newC == COLS || grid[newR][newC] != INF) continue;

                grid[newR][newC] = grid[row][col] + 1;
                q.push({newR, newC});
            }
        }
    }
};
